src/lib.rs

```diff
@@ -1034,20 +1035,46 @@ impl<T: Clone + Integer + Signed> Signed for Ratio<T> {
 }
 
 // String conversions
-impl<T> fmt::Display for Ratio<T>
-where
-    T: fmt::Display + Eq + One,
-{
-    /// Renders as `numer/denom`. If denom=1, renders as numer.
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        if self.denom.is_one() {
-            write!(f, "{}", self.numer)
-        } else {
-            write!(f, "{}/{}", self.numer, self.denom)
+macro_rules! impl_formatting {
+    ($fmt_trait:ident, $prefix:expr, $fmt_str:expr, $fmt_alt:expr) => {
+        impl<T: $fmt_trait + PartialEq + One> $fmt_trait for Ratio<T> {
```

Comment:
I think I'd prefer to have the formatting traits require `T: Clone + Integer`, like most of the API. Even if we don't *need* it now, it's a slight bit of future-proofing to have those constraints available.

Reply:
That would let you restore better handling for negative values too, right?

Reply:
By better handling of negatives, do you mean like in 
`assert_eq!(&format!("{:#b}", -half_i8), "0b11111111/0b10");`? If so, what should `&format!("{:#b}"` equal? I feel like this is how Binary formatting should work.

Reply:
Adding these trait requirements breaks my code that tests `Ratio`s of `f32`s. Maybe I never should have written that test code, but how should I test that Ratio works with the *Exp formatting traits?

Reply:
In your first few commits, you had `non_negative` checking for the sign placement, which seems to be gone now. That's what I'm referring to for better handling of negative values. I'm still unsure how I feel about that binary formatting...

I think it's ok to lose `f32` support, just test Exp on newer toolchains where supported for integers.

Reply:
> In your first few commits, you had non_negative checking for the sign placement

I removed this for a reason - this line was problematic `f.pad_integral(non_negative, $prefix, &pre_pad[1..])`, which fails on `assert_eq!(&format!("{:#b}", -half_i8), "0b11111111/0b10");`, as it removes the first `1` from the numerator, `11111111`. More generally that line breaks binary (and octal and hex) formatting of any negative number. 

I thought I fixed this problem, by removing the negative check, but in doing so I created another problem. The current code fails on `assert_fmt_eq!(format_args!("{:+}", -_2), "-2");` (a new test case) with `"+-2"`.

As a quick fix for both these problems, I came up with
```rust
if stringify!($fmt_trait) == "Display" {
    let non_negative = *self >= Self::zero();
    if non_negative {
        f.pad_integral(non_negative, $prefix, &pre_pad)
    } else {
        f.pad_integral(non_negative, $prefix, &pre_pad[1..])
    }
} else {
    f.pad_integral(true, $prefix, &pre_pad)
}
```
Which passes all test cases, but is hacky, and not just because of `stringify!`, which could be replaced by a macro parameter. It is hacky because it depends on the fact that `Display`ing a negative number starts with `-`, which is not clearly documented in the `Display` trait, and may not be true on a user type implementing `Integer`. `Integer` seems to corresponds to a mathematical ring, in which negative elements may not be distinct from positive ones (i.e. the ring of integers modulo 4 in which -1 == 3). I realize this is a stretch, but it could happen. I could of course avoid this trouble by not relying on `pad_integral`, and rolling my own solution. All I really need is a `pad_integral` where I can override a few settings in the formatter, or a format string that specifically ignores sign.

Reply:
Maybe for this particular case, `self.numer < T::zero() && f.sign_plus()`, we could negate the numerator before formatting it at all. Then `pad_integral` would completely own the responsibility for any sign output.

> `Integer` seems to corresponds to a mathematical ring, in which negative elements may not be distinct from positive ones (i.e. the ring of integers modulo 4 in which -1 == 3).

I'm not sure I understand this scenario though -- do negative elements really exist in this case? i.e. would they compare less than zero at all? You seem to be supposing a type that would track negative state somehow, but still prints in modulo, as a positive value.

Anyway, `trim_left('-')` would be a bit less hacky, or maybe `strip_prefix('-')` when that's stabilized.

Reply:
>  Maybe for this particular case, self.numer < T::zero() && f.sign_plus(), we could negate the numerator before formatting it at all

I didn't want to negate because of the overflow issue - `-i32::min_value() == i32::min_value()` i.e. #6 . But that looks like the best option.

> I'm not sure I understand this scenario though -- do negative elements really exist in this case?

In the strictest sense, in the ring of integers mod 4 there is no positive or negative. In a less strict sense, all the elements are positive and correspond to 0,1,2,3. There is a negation operator (multiplication by -1), but its weird, since -1 is actually 3. I was getting rant-y though. The larger point was that I didn't want to do string manipulation, since it is fragile. I think I'm just going to negate the numerator, as you suggested.

Reply:
> I didn't want to negate because of the overflow issue - `-i32::min_value() == i32::min_value()` i.e. #6 . But that looks like the best option.

Ugh, not being able to print `min_value()` would be sad.

Reply:
I decided to do the stringy way in the end. Honestly the logic got a bit complicated when negating. 

We're making the assumption that no integer type of positive value is printed as starting with '-' unless it is truly negative. That's an assumption I'm willing to make. Should it be documented anywhere in particular, or is it pretty obvious?

Reply:
(But no_std doesn't have the luxury of string manipulations, so no_std uses negation)